### DISCUSSION TITLE
Extract Nodal State Space Matrices under Krylov ROM

### DISCUSSION THREAD

**Discussion by xxyao-95** [ImperialCollegeLondon/sharpy#154] [Q&A] [answered]
Hi,

I am using LinearAssembler to extract the State Space ABCD matrices with Krylov ROM.

I want to extract the state space in nodal coordinates so that I could manipulate the output y  for a LQG controller.

The problem that I am having now is that Sharpy freezes during ROM, here is what console show when it stops:
```
Scaling beam according to reduced time...
	Setting the beam time step to (0.1250)
Updating C and K matrices and natural frequencies with new normalised time...
Model Order Reduction in progress...
Moment Matching Krylov Model Reduction
	Construction Algorithm:
		mimo_rational_arnoldi
	Interpolation points:
		sigma = 0.000000 + 0.000000j [rad/s]

	Krylov order:
		r = 6
	Constructing controllability space
```

I think there is something wrong with my settings for LinearAssember:
```
{'linear_system': 'LinearAeroelastic', 
'inout_coords': 'nodes', 
'linear_system_settings': {'rigid_body_motion': False, 
'beam_settings': {'discrete_time': 'on', 
                            'newmark_damp': 5e-05, 
                            'discr_method': 'newmark', 
                            'dt': 0.1143, 'proj_modes': 'undamped', 
                            'use_euler': 'off', 'num_modes': 4, 
                            'print_info': 'on', 'gravity': 'on', 
                            'remove_sym_modes': 'on', 'remove_dofs': []}, 
'aero_settings': {'dt': 0.1143, 
                          'integr_order': 2, 
                          'density': 1.02, 
                          'remove_predictor': 'on', 
                           'use_sparse': 'on', 
                           'rigid_body_motion': 'off', 
                           'use_euler': 'off', 
                           'remove_inputs': [''], 
                           'gust_assembler': 'leading_edge', 
                           'rom_method': ['Krylov'], 
                           'ScalingDict': {'length': 0.9144, 'speed': 1.0, 'density': 1.02}, 
 'rom_method_settings': {'Krylov': {'algorithm': 'mimo_rational_arnoldi', 'r': 6, 'frequency': array([0.])}}}}}
```
Could you let me know what went wrong with my LinearAssembler Settings?

Mod edit: changed block code to use the Markdown format (3 x `)

**Comment by ngoiz** [comment, marked as answer]
This is likely because your structural subsystem is assembled in nodal coordinates and the ROM is dealing with a huge input/output space. What you need is your system to be in modal space (i.e. low inputs/outputs so the ROM works) and then once you have that you manipulate the assembled systems' inputs and outputs to go back to the nodal space.

Adding to your `beam_settings` above: 
```
                                          'beam_settings': {'modal_projection': 'on',
                                                            'inout_coords': 'modes'}
```
should make it work.

Note, your `LinearAssembler` setting `inout_coords` should be `inout_coordinates` and that should be equal to `nodes` as you have it. Careful with those because a typo would mean SHARPy ignores them (this will change in the next version though)

**Reply by xxyao-95** [reply]
Thank you so much! I run Sharpy with the comments as discussed and I found out Y now has 576 states and x has 41 states which means x is in modal coordinates and Y is in nodal coordinates, hence I can manipulate the C matrix to find the desired output.

Thank you so much for your help!
